pkg/pushsync/pushsync_test.go

```diff
@@ -2,4 +2,336 @@
+		Storer:        storer,
+	})
+	defer ps.Close()
+
+	// TODO: this needs to go away once we have a simpler recorder
+	recorder.SetProtocols(ps.Protocol())
+
+	stream, err := recorder.NewStream(context.Background(), pivotNode, nil, pushsync.ProtocolName, pushsync.ProtocolVersion, pushsync.StreamName)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer stream.Close()
+	w, _ := protobuf.NewWriterAndReader(stream)
+
+	// this triggers the handler of the pivot with a delivery stream
+	err = w.WriteMsg(&pb.Delivery{
+		Address: chunkAddress.Bytes(),
+		Data:    chunkData,
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	for i := 0; i < 20; i++ {
+		mtx.Lock()
+		called := targetCalled
+		mtx.Unlock()
+
+		if !called {
+			time.Sleep(10 * time.Millisecond)
+		}
+	}
+
+	records, err := recorder.Records(closestPeer, pushsync.ProtocolName, pushsync.ProtocolVersion, pushsync.StreamName)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if l := len(records); l != 1 {
+		t.Fatalf("got %v records, want %v", l, 1)
+	}
+}
+
+// TestNoForwardChunk tests that the closest node to a chunk doesn't forward it to other nodes.
+func TestNoForwardChunk(t *testing.T) {
+	logger := logging.New(ioutil.Discard, 0)
+
+	// chunk data to upload
+	chunkAddress := swarm.MustParseHexAddress("7000000000000000000000000000000000000000000000000000000000000000") // binary 0111
+	chunkData := []byte("1234")
+
+	// create a pivot node and a cluster of nodes
+	pivotNode := swarm.MustParseHexAddress("6000000000000000000000000000000000000000000000000000000000000000") // pivot is 0110
+	connectedPeers := []p2p.Peer{
+		{
+			Address: swarm.MustParseHexAddress("8000000000000000000000000000000000000000000000000000000000000000"), // binary 1000 -> po 0
+		},
+		{
+			Address: swarm.MustParseHexAddress("4000000000000000000000000000000000000000000000000000000000000000"), // binary 0100 -> po 1
+		},
+		{
+			Address: swarm.MustParseHexAddress("0000000000000000000000000000000000000000000000000000000000000000"), // binary 0000 -> po 0
+		},
+	}
+
+	// mock a connectivity between the nodes
+	p2ps := p2pmock.New(p2pmock.WithConnectFunc(func(ctx context.Context, addr ma.Multiaddr) (swarm.Address, error) {
+		return pivotNode, nil
+	}), p2pmock.WithPeersFunc(func() []p2p.Peer {
+		return connectedPeers
+	}))
+
+	// Create a full connectivity driver
+	fullDriver := full.New(mock.NewDiscovery(), addressbook.New(mockstate.NewStateStore()), p2ps, logger, pivotNode)
+	storer, err := localstore.New("", pivotNode.Bytes(), nil, logger)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// how many calls to the handler we've handled
+	calls := 0
+
+	// setup the stream recorder to record stream data
+	recorder := streamtest.New(
+		streamtest.WithMiddlewares(func(f p2p.HandlerFunc) p2p.HandlerFunc {
+			calls++
+			if runtime.GOOS == "windows" {
+				// windows has a bit lower time resolution
+				// so, slow down the handler with a middleware
+				// not to get 0s for rtt value
+				time.Sleep(100 * time.Millisecond)
```

Comment:
`ewww^2`

Reply:
I think that this is from the pingpong test and do not think that it is needed here.